packages/core/src/subscriber/chainSubscribers.ts

```diff
@@ -0,0 +1,181 @@
+import {
+    services,
+    utils,
+    config,
+    contracts,
+    database,
+} from '../../';
+import { ethers } from 'ethers';
+
+class ChainSubscribers {
+    provider: ethers.providers.WebSocketProvider;
+    ifaceCommunityAdmin: ethers.utils.Interface;
+    filterTopics: string[][];
+
+    constructor(provider: ethers.providers.WebSocketProvider) {
+        this.provider = provider;
+        this.ifaceCommunityAdmin = new ethers.utils.Interface(
+            contracts.CommunityAdminABI
+        );
+
+        this.filterTopics = [
+            [
+                ethers.utils.id(
+                    'CommunityAdded(address,address[],uint256,uint256,uint256,uint256,uint256,uint256,uint256)'
+                ),
+                ethers.utils.id('CommunityRemoved(address)'),
+            ],
+        ];
+        this.recover();
+    }
+
+    stop() {
+        this.provider.removeAllListeners();
+        services.app.ImMetadataService.setRecoverBlockUsingLastBlock();
+    }
+
+    recover() {
+        this._setupListener(this.provider);
+        // we start the listener alongside with the recover system
+        // so we know we don't lose events.
+        services.app.ImMetadataService.getRecoverBlock().then((block) =>
+            this._runRecoveryTxs(block, this.provider).then(() =>
+                services.app.ImMetadataService.removeRecoverBlock()
+            )
+        );
+    }
+
+    async _runRecoveryTxs(
+        startFromBlock: number,
+        provider: ethers.providers.JsonRpcProvider
+    ) {
+        utils.Logger.info('Recovering past events...');
+        const rawLogs = await provider.getLogs({
+            fromBlock: startFromBlock,
+            toBlock: 'latest',
+            topics: this.filterTopics,
+        });
```

Comment:
There's a possible problem here. The provider is a websocket.
I'd suggest a slight change. On the constructor, don't pass any parameters and instantiate the provider within the class. Read from env and create a websocket with websocket url and another provider with http just to use this getLogs. Or maybe, because it's used outside, pass both the http and websocket